src/Algebra/Module.agda

```diff
@@ -0,0 +1,283 @@
+    0ᴹ : Carrierᴹ
+    isSemimodule : IsSemimodule commutativeSemiring _+ᴹ_ _*ₗ_ 0ᴹ
+
+  open IsSemimodule isSemimodule public
+
+  private
+    module L = LFP _≈_ _≈ᴹ_
+    module R = RFP _≈_ _≈ᴹ_
+
+  leftSemimodule : LeftSemimodule semiring m ℓm
+  leftSemimodule = record { isLeftSemimodule = isLeftSemimodule }
+
+  open LeftSemimodule leftSemimodule public
+    using ( +ᴹ-commutativeMonoid; +ᴹ-monoid; +ᴹ-semigroup; +ᴹ-magma
+          ; +ᴹ-rawMonoid; +ᴹ-rawMagma)
+
+  _*ᵣ_ : Opᵣ Carrier Carrierᴹ
+  _*ᵣ_ = flip _*ₗ_
+
+  *ₗ-comm : L.Commutative _*ₗ_
+  *ₗ-comm x y m =
+    M-trans (M-sym (*ₗ-assoc x y m))
+            (M-trans (*ₗ-cong (*-comm _ _) M-refl)
+                     (*ₗ-assoc y x m))
+
+  *ᵣ-comm : R.Commutative _*ᵣ_
+  *ᵣ-comm m x y = *ₗ-comm y x m
+
+  rightSemimodule : RightSemimodule semiring m ℓm
+  rightSemimodule = record
+    { Carrierᴹ = Carrierᴹ
+    ; _≈ᴹ_ = _≈ᴹ_
+    ; _+ᴹ_ = _+ᴹ_
+    ; _*ᵣ_ = _*ᵣ_
+    ; 0ᴹ = 0ᴹ
+    ; isRightSemimodule = record
+      { +ᴹ-isCommutativeMonoid = +ᴹ-isCommutativeMonoid
+      ; *ᵣ-cong = flip *ₗ-cong
+      ; *ᵣ-zeroʳ = *ₗ-zeroˡ
+      ; *ᵣ-distribˡ = λ m x y → *ₗ-distribʳ m x y
+      ; *ᵣ-identityʳ = λ m → *ₗ-identityˡ m
+      ; *ᵣ-assoc = λ m x y → M-trans (*ₗ-comm y x m) (M-sym (*ₗ-assoc x y m))
+      ; *ᵣ-zeroˡ = *ₗ-zeroʳ
+      ; *ᵣ-distribʳ = λ x m n → *ₗ-distribˡ x m n
+      }
+    }
+
+  open RightSemimodule rightSemimodule public
+    using ( *ᵣ-cong; *ᵣ-zeroʳ; *ᵣ-distribˡ; *ᵣ-identityʳ; *ᵣ-assoc; *ᵣ-zeroˡ
+          ; *ᵣ-distribʳ; isRightSemimodule)
+
+record Module {r ℓr} (commutativeRing : CommutativeRing r ℓr) m ℓm
+              : Set (r ⊔ ℓr ⊔ suc (m ⊔ ℓm)) where
+  open CommutativeRing commutativeRing
+
+  infixr 8 -ᴹ_
+  infixr 7 _*ₗ_
+  infixl 6 _+ᴹ_
+  infix 4 _≈ᴹ_
+
+  field
+    Carrierᴹ : Set m
+    _≈ᴹ_ : Rel Carrierᴹ ℓm
+  open Str _≈ᴹ_
+  field
+    _+ᴹ_ : Op₂ Carrierᴹ
+    _*ₗ_ : Opₗ Carrier Carrierᴹ
+    0ᴹ : Carrierᴹ
+    -ᴹ_ : Op₁ Carrierᴹ
+    isModule : IsModule commutativeRing _+ᴹ_ _*ₗ_ 0ᴹ -ᴹ_
+
+  open IsModule isModule public
+
+  semimodule : Semimodule commutativeSemiring m ℓm
+  semimodule = record
+    { _≈ᴹ_ = _≈ᴹ_
+    ; _+ᴹ_ = _+ᴹ_
+    ; _*ₗ_ = _*ₗ_
+    ; 0ᴹ = 0ᴹ
+    ; isSemimodule = isSemimodule
+    }
+
+  open Semimodule semimodule public
+    using ( leftSemimodule; isLeftSemimodule; _*ᵣ_; *ₗ-comm; *ᵣ-comm
+          ; rightSemimodule; isRightSemimodule; *ᵣ-cong; *ᵣ-zeroʳ; *ᵣ-distribˡ
+          ; *ᵣ-identityʳ; *ᵣ-assoc; *ᵣ-zeroˡ; *ᵣ-distribʳ
+          ; +ᴹ-commutativeMonoid; +ᴹ-monoid; +ᴹ-semigroup; +ᴹ-magma
+          ; +ᴹ-rawMonoid; +ᴹ-rawMagma)
+
+  leftModule : LeftModule ring m ℓm
+  leftModule = record { isLeftModule = isLeftModule }
+
+  open LeftModule leftModule public
+    using (+ᴹ-abelianGroup; +ᴹ-group)
+
+  isRightModule : IsRightModule ring _+ᴹ_ _*ᵣ_ 0ᴹ -ᴹ_
+  isRightModule = record
+    { isRightSemimodule = record
+      { +ᴹ-isCommutativeMonoid = +ᴹ-isCommutativeMonoid
+      ; *ᵣ-cong = *ᵣ-cong
```

Comment:
Same here: probably worth lifting the `isRightSemimodule` proof to the toplevel.